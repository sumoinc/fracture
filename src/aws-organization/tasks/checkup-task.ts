import { Component, Task } from "projen";
import { CdkProject } from "../../cdk";
import { localOnlyOptions } from "../../tasks/options";

export class CheckupTask extends Component {
  public static SCRIPTNAME = "checkup";

  /**
   * The task
   */
  readonly task: Task;

  constructor(public readonly project: CdkProject) {
    super(project);

    this.task = project.addTask(CheckupTask.SCRIPTNAME, {
      description: `Runs everythiung but sso-login to make sure all environments are healthy and bootstrapped properly.`,
      ...localOnlyOptions,
    });
  }
}
