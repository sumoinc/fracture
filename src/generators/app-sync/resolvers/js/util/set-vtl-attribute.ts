import { ResourceAttributeGenerator } from "../../../../../services/resource-attribute";
import { StructureAttribute } from "../../../../../services/structure-attribute";
import { TypescriptStrategy } from "../../../../ts/strategy";

export const setVtlAttribute = (attribute: StructureAttribute) => {
  const strategy = TypescriptStrategy.of(attribute.project);
  const generator = attribute.generator;
  const name = strategy.formatAttributeName(attribute.name);
  const shortName = strategy.formatAttributeName(attribute.shortName);

  // non-generated
  if (generator === ResourceAttributeGenerator.NONE) {
    return `"${shortName}": $util.dynamodb.toDynamoDBJson($ctx.args.input.${name}),`;
    // generated values
  } else {
    switch (generator) {
      case ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP:
        return `"${shortName}": "$util.time.nowISO8601()"`;
      case ResourceAttributeGenerator.GUID:
        return `"${shortName}": "$util.autoId()"`;
      case ResourceAttributeGenerator.TYPE:
        return `"${shortName}": "unknown"`;

      default:
        throw new Error(`Unknown generator: ${generator}`);
    }
  }
};
