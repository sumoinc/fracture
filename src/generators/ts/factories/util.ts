import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
