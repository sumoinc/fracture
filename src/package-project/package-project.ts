import { Project } from "projen";
import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { Jest } from "../tests/jest";

export interface PackageProjectOptions
  extends Omit<TypeScriptProjectOptions, "defaultReleaseBranch"> {
  // readonly parent: MonorepoProject;
  // readonly testingOptions?: TestingOptions;
}

/**
 * Package Project
 */
export class PackageProject extends TypeScriptProject {
  /**
   * Returns all packages in the monorepo.
   */
  public static all(project: Project): Array<PackageProject> {
    const isDefined = (p: Project): p is PackageProject =>
      p instanceof PackageProject;
    return project.root.subprojects.filter(isDefined);
  }

  /**
   * The testing settings for this project.
   */
  // public testing?: Testing;

  constructor(options: PackageProjectOptions) {
    /**
     * Is this a root level project? This determines what folder structure to
     * use when storing the project.
     */
    //const isRootProject: boolean = options.parent === undefined;

    const outdir: string = options.parent ? `packages/${options.name}` : "";

    const artifactsDirectory: string = options.parent ? "dist" : "dist";

    /**
     * Builds a cachable output path at the root of the monorepo for dist
     * and other deployable artifacts.
     */
    /*
    const buildOutputPath = (name: string) => {
      return join(
        "../..",
        options.parent.deploymentArtifactRoot,
        `packages/${options.name}`,
        name
      );
    };*/

    super({
      /*************************************************************************
       * INPUTS
       ************************************************************************/

      ...options,

      /*************************************************************************
       * FORCED OPTIONS
       ************************************************************************/

      /**
       * Use main branch. It's 2023 by the way.
       */
      defaultReleaseBranch: "main",

      /**
       * Enable prettier for some great formatting.
       */
      prettier: true,

      /**
       * PNPM 8 all the way baby!
       */
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: "8",

      /**
       * No license for now.
       */
      licensed: false,

      /**
       * Turn off the native jest support, we will be adding our own Jest
       * config beflow.
       */
      jest: false,

      /**
       * All package projects are placed into the packages path structure.
       */
      // outdir: `packages/${options.name}`,
      outdir,

      /**
       * Always package things in packages folder
       */
      package: true,

      /**
       * Final distributable assets go to a common directory in the root.
       */
      // artifactsDirectory: buildOutputPath("dist"),
      artifactsDirectory,
    });

    /***************************************************************************
     *
     * PROJEN SETUP
     *
     **************************************************************************/

    [
      ".DS_Store",
      ".gitattributes",
      ".prettierignore",
      ".prettierrc.json",
    ].forEach((f) => {
      this.addPackageIgnore(f);
    });
    this.addGitIgnore(".DS_Store");

    /***************************************************************************
     *
     * SETUP TESTING
     *
     **************************************************************************/

    new Jest(this);

    /***************************************************************************
     *
     * CONFIGURE TURBO
     *
     **************************************************************************/

    // if (Turbo.isActive(this)) {
    //   const turbo = Turbo.of(this)!;

    //   /**
    //    * Testing uses ts-jest so it doesn't depend on anyother tasks.
    //    */
    //   turbo.addTask(this, {
    //     projenTask: this.testTask,
    //     cachable: true,
    //     inputPaths: ["jest.config.json", ...Turbo.commonInputs(this)],
    //     outputPaths: ["coverage/**", "test-reports/**"],
    //   });

    //   /**
    //    * Compile and Package need to happen in a particular order.
    //    */
    //   const turboCompile = turbo.addTask(this, {
    //     projenTask: this.compileTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.libdir}/**`],
    //   });
    //   const turboPackage = turbo.addTask(this, {
    //     projenTask: this.packageTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.artifactsDirectory}/**`],
    //   });
    //   turboCompile.thenRun(turboPackage);
    // }
  }
}
