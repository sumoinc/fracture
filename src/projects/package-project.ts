import { Project } from "projen";
import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { commonProjectSetup } from "./common";
import { Jest } from "../tests/jest";

export interface PackageProjectOptions
  extends Omit<TypeScriptProjectOptions, "defaultReleaseBranch"> {
  // readonly parent: MonorepoProject;
  // readonly testingOptions?: TestingOptions;
}

/**
 * Package Project
 */
export class PackageProject extends TypeScriptProject {
  /**
   * Returns all packages in the monorepo.
   */
  public static all(project: Project): Array<PackageProject> {
    const isDefined = (p: Project): p is PackageProject =>
      p instanceof PackageProject;
    return project.root.subprojects.filter(isDefined);
  }

  /**
   * The testing settings for this project.
   */
  // public testing?: Testing;

  constructor(options: PackageProjectOptions) {
    /**
     * Is this a root level project? This determines what folder structure to
     * use when storing the project.
     */
    //const isRootProject: boolean = options.parent === undefined;

    const outdir: string = options.parent ? `packages/${options.name}` : "";

    const artifactsDirectory: string = options.parent ? "dist" : "dist";

    /**
     * Builds a cachable output path at the root of the monorepo for dist
     * and other deployable artifacts.
     */
    /*
    const buildOutputPath = (name: string) => {
      return join(
        "../..",
        options.parent.deploymentArtifactRoot,
        `packages/${options.name}`,
        name
      );
    };*/

    super({
      /*************************************************************************
       * DEFAULTS
       ************************************************************************/

      /**
       * No license by default.
       */
      licensed: false,

      /**
       * Use Node 18.x by default for package builds
       */
      workflowNodeVersion: "18",

      /**
       * Use main branch. It's 2023 by the way.
       */
      defaultReleaseBranch: "main",

      /**
       * Enable prettier for some great formatting.
       */
      prettier: true,

      /**
       * PNPM 8 all the way baby!
       */
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: "8",

      /**
       * Turn off the native jest support, we will be adding our own Jest
       * config beflow.
       */
      jest: false,

      /**
       * All package projects are placed into the packages path structure.
       */
      // outdir: `packages/${options.name}`,
      outdir,

      /**
       * Always package assets for distribution.
       */
      package: true,

      /**
       * Final distributable assets go to a common directory in the root.
       */
      // artifactsDirectory: buildOutputPath("dist"),
      artifactsDirectory,

      /*************************************************************************
       * INPUTS
       ************************************************************************/

      ...options,
    });

    /***************************************************************************
     *
     * COMMON SETUP
     *
     **************************************************************************/

    commonProjectSetup(this);

    /***************************************************************************
     *
     * SETUP IGNORES
     *
     **************************************************************************/

    [
      ".DS_Store",
      ".gitattributes",
      ".prettierignore",
      ".prettierrc.json",
      ".projenrc.ts",
      "node_modules",
      "/**/*.spec.*",
    ].forEach((f) => {
      this.addPackageIgnore(f);
    });

    [".DS_Store"].forEach((f) => {
      this.addGitIgnore(f);
    });

    /***************************************************************************
     *
     * SETUP TESTING
     *
     **************************************************************************/

    new Jest(this);

    /***************************************************************************
     *
     * CONFIGURE TURBO
     *
     **************************************************************************/

    // if (Turbo.isActive(this)) {
    //   const turbo = Turbo.of(this)!;

    //   /**
    //    * Testing uses ts-jest so it doesn't depend on anyother tasks.
    //    */
    //   turbo.addTask(this, {
    //     projenTask: this.testTask,
    //     cachable: true,
    //     inputPaths: ["jest.config.json", ...Turbo.commonInputs(this)],
    //     outputPaths: ["coverage/**", "test-reports/**"],
    //   });

    //   /**
    //    * Compile and Package need to happen in a particular order.
    //    */
    //   const turboCompile = turbo.addTask(this, {
    //     projenTask: this.compileTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.libdir}/**`],
    //   });
    //   const turboPackage = turbo.addTask(this, {
    //     projenTask: this.packageTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.artifactsDirectory}/**`],
    //   });
    //   turboCompile.thenRun(turboPackage);
    // }
  }
}
