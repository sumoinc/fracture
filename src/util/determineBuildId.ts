import { execSync } from "child_process";

/**
 * Returns the current git branch name, without folder.
 *
 * ie: feature/1234 returns 1234
 *
 * @returns {string}
 */
export const determineBuildId = () => {
  const branchName = execSync("git rev-parse --abbrev-ref HEAD")
    .toString("utf8")
    .replace(/[\n\r\s]+$/, "");
  const branchPath = branchName.split("/");
  return branchPath[branchPath.length - 1];
};
