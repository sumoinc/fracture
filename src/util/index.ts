export * from "./determineBuildId";
export * from "./test-util";
