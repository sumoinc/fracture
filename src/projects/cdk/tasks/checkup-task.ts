import { Component, Task } from "projen";
import { TypeScriptProject } from "projen/lib/typescript";
import { BootstrapTask } from "./bootstrap-task";
import { ProfileTask } from "./profile-task";
import { findOrCreateTask, localOnlyTaskCondition } from "../../../tasks/task-utils";

export class CheckupTask extends Component {
  public static SCRIPTNAME = "checkup";

  /**
   * The task
   */
  readonly task: Task;

  constructor(public readonly project: TypeScriptProject) {
    super(project);

    this.task = project.addTask(CheckupTask.SCRIPTNAME, {
      description: `Runs everythiung but sso-login to make sure all environments are healthy and bootstrapped properly.`,
      ...localOnlyTaskCondition,
    });

    // this.task.exec(`rm -rf ${this.deploymentArtifactRoot}`);

    // make sure all profiles are created locally
    this.task.spawn(findOrCreateTask(this.project, ProfileTask.SCRIPTNAMEBASE));

    // make sure all environments are properly bootstrapped
    this.task.spawn(
      findOrCreateTask(this.project, BootstrapTask.SCRIPTNAMEBASE)
    );
  }
}
