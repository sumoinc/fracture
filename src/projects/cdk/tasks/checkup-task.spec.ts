import { CheckupTask } from "./checkup-task";
import { synthFiles } from "../../../util";
import { PackageProject } from "../../package-project";

describe("Success conditions", () => {
  test("Smoke Test", () => {
    const project = new PackageProject({
      name: "my-project",
    });

    // build task based on bootstrap config
    const task = new CheckupTask(project);

    expect(task).toBeTruthy();
  });
});

describe("Files", () => {
  test(".projen/tasks.json", () => {
    const project = new PackageProject({
      name: "my-project",
    });

    // build task
    new CheckupTask(project);

    const content = synthFiles(project);
    expect(content[".projen/tasks.json"]).toBeTruthy();
    expect(content[".projen/tasks.json"]).toMatchSnapshot();

    // console.log(JSON.stringify(content[".projen/tasks.json"], null, 2));
  });
});
