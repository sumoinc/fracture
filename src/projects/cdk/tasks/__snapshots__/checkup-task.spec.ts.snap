// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Files .projen/tasks.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".",
  "env": Object {
    "PATH": "$(pnpm -c exec \\"node --print process.env.PATH\\")",
  },
  "tasks": Object {
    "bootstrap": Object {
      "name": "bootstrap",
    },
    "build": Object {
      "description": "Full release build",
      "name": "build",
      "steps": Array [
        Object {
          "spawn": "default",
        },
        Object {
          "spawn": "pre-compile",
        },
        Object {
          "spawn": "compile",
        },
        Object {
          "spawn": "post-compile",
        },
        Object {
          "spawn": "test",
        },
        Object {
          "spawn": "package",
        },
      ],
    },
    "bump": Object {
      "condition": "! git log --oneline -1 | grep -q \\"chore(release):\\"",
      "description": "Bumps version based on latest git tag and generates a changelog entry",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "bump",
      "steps": Array [
        Object {
          "builtin": "release/bump-version",
        },
      ],
    },
    "checkup": Object {
      "condition": "[ ! -n \\"$CI\\" ]",
      "description": "Runs everythiung but sso-login to make sure all environments are healthy and bootstrapped properly.",
      "name": "checkup",
      "steps": Array [
        Object {
          "spawn": "profiles",
        },
        Object {
          "spawn": "bootstrap",
        },
      ],
    },
    "clobber": Object {
      "condition": "git diff --exit-code > /dev/null",
      "description": "hard resets to HEAD of origin and cleans the local repo",
      "env": Object {
        "BRANCH": "$(git branch --show-current)",
      },
      "name": "clobber",
      "steps": Array [
        Object {
          "exec": "git checkout -b scratch",
          "name": "save current HEAD in \\"scratch\\" branch",
        },
        Object {
          "exec": "git checkout $BRANCH",
        },
        Object {
          "exec": "git fetch origin",
          "name": "fetch latest changes from origin",
        },
        Object {
          "exec": "git reset --hard origin/$BRANCH",
          "name": "hard reset to origin commit",
        },
        Object {
          "exec": "git clean -fdx",
          "name": "clean all untracked files",
        },
        Object {
          "say": "ready to rock! (unpushed commits are under the \\"scratch\\" branch)",
        },
      ],
    },
    "compile": Object {
      "description": "Only compile",
      "name": "compile",
      "steps": Array [
        Object {
          "exec": "tsc --build",
        },
      ],
    },
    "default": Object {
      "description": "Synthesize project files",
      "name": "default",
      "steps": Array [
        Object {
          "exec": "ts-node --project tsconfig.dev.json .projenrc.ts",
        },
      ],
    },
    "eject": Object {
      "description": "Remove projen from the project",
      "env": Object {
        "PROJEN_EJECTING": "true",
      },
      "name": "eject",
      "steps": Array [
        Object {
          "spawn": "default",
        },
      ],
    },
    "eslint": Object {
      "description": "Runs eslint against the codebase",
      "name": "eslint",
      "steps": Array [
        Object {
          "exec": "eslint --ext .ts,.tsx --fix --no-error-on-unmatched-pattern $@ src test build-tools projenrc .projenrc.ts",
          "receiveArgs": true,
        },
      ],
    },
    "install": Object {
      "description": "Install project dependencies and update lockfile (non-frozen)",
      "name": "install",
      "steps": Array [
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
      ],
    },
    "install:ci": Object {
      "description": "Install project dependencies using frozen lockfile",
      "name": "install:ci",
      "steps": Array [
        Object {
          "exec": "pnpm i --frozen-lockfile",
        },
      ],
    },
    "package": Object {
      "description": "Creates the distribution package",
      "name": "package",
      "steps": Array [
        Object {
          "exec": "mkdir -p dist/js",
        },
        Object {
          "exec": "mv $(pnpm pack) dist/js/",
        },
      ],
    },
    "post-compile": Object {
      "description": "Runs after successful compilation",
      "name": "post-compile",
    },
    "post-upgrade": Object {
      "description": "Runs after upgrading dependencies",
      "name": "post-upgrade",
    },
    "pre-compile": Object {
      "description": "Prepare the project for compilation",
      "name": "pre-compile",
    },
    "profiles": Object {
      "name": "profiles",
    },
    "release": Object {
      "description": "Prepare a release from \\"main\\" branch",
      "env": Object {
        "RELEASE": "true",
      },
      "name": "release",
      "steps": Array [
        Object {
          "exec": "rm -fr dist",
        },
        Object {
          "spawn": "bump",
        },
        Object {
          "spawn": "build",
        },
        Object {
          "spawn": "unbump",
        },
        Object {
          "exec": "git diff --ignore-space-at-eol --exit-code",
        },
      ],
    },
    "test": Object {
      "description": "Run tests",
      "name": "test",
      "steps": Array [
        Object {
          "spawn": "eslint",
        },
        Object {
          "exec": "jest --passWithNoTests --updateSnapshot",
          "receiveArgs": true,
        },
      ],
    },
    "test:watch": Object {
      "description": "Run jest in watch mode",
      "name": "test:watch",
      "steps": Array [
        Object {
          "exec": "jest --watch",
        },
      ],
    },
    "unbump": Object {
      "description": "Restores version to 0.0.0",
      "env": Object {
        "BUMPFILE": "dist/version.txt",
        "CHANGELOG": "dist/changelog.md",
        "OUTFILE": "package.json",
        "RELEASETAG": "dist/releasetag.txt",
        "RELEASE_TAG_PREFIX": "",
      },
      "name": "unbump",
      "steps": Array [
        Object {
          "builtin": "release/reset-version",
        },
      ],
    },
    "upgrade": Object {
      "description": "upgrade dependencies",
      "env": Object {
        "CI": "0",
      },
      "name": "upgrade",
      "steps": Array [
        Object {
          "exec": "pnpx npm-check-updates@16 --upgrade --target=minor --peer --dep=dev,peer,prod,optional --filter=@types/jest,@types/node,@typescript-eslint/eslint-plugin,@typescript-eslint/parser,constructs,eslint-config-prettier,eslint-import-resolver-typescript,eslint-plugin-import,eslint-plugin-prettier,eslint,jest,jest-junit,prettier,projen,standard-version,ts-jest,ts-node,typescript",
        },
        Object {
          "exec": "pnpm i --no-frozen-lockfile",
        },
        Object {
          "exec": "pnpm update @types/jest @types/node @typescript-eslint/eslint-plugin @typescript-eslint/parser constructs eslint-config-prettier eslint-import-resolver-typescript eslint-plugin-import eslint-plugin-prettier eslint jest jest-junit prettier projen standard-version ts-jest ts-node typescript",
        },
        Object {
          "exec": "npx projen",
        },
        Object {
          "spawn": "post-upgrade",
        },
      ],
    },
    "watch": Object {
      "description": "Watch & compile in the background",
      "name": "watch",
      "steps": Array [
        Object {
          "exec": "tsc --build -w",
        },
      ],
    },
  },
}
`;
